share/completions/iwctl.fish

```diff
@@ -0,0 +1,138 @@
+# Execute an `iwctl ... list` command and parse output
+function __iwctl_filter
+    # set results "iwctl $cmd list | tail -n +5"
+    # if test -n "$empty"
+    #     set -a results "| string match --invert '*$empty*'"
+    # end
+    # eval "$results" | awk '{print $2}'
+    # awk does not work on multiline entries, therefor we use string match,
+    # which has the added benefit of filtering out the `No devices in ...` lines
+
+    # remove color escape sequences
+    set results (iwctl $argv | string replace -ra '\e\[[\d;]+m' '')
+    # calculate column widths
+    set headers $results[3]
+    set header_spaces (string match -a -r "  +" $headers | string length)
+    set header_non_spaces (string match -a -r "[^ ]+" $headers | string length)
+
+    # only take lines starting with `  `, i.e., no `No devices ...`
+    # then take the first column as substring
+    echo $results[5..]\n | string match -r "  .*" | string sub -s $header_spaces[1] -l (math $header_non_spaces[1] + $header_spaces[2]) | string trim
```

Comment:
well I want to map each item individually, so I thought I need new lines to separate. Maybe there is a better way to map lists I don't know.

Reply:
You can just give the arguments to string:

```fish
string match -r "  .*" -- $results[5..] | string sub...
```

Alternatively, if you want to get all lines starting with X spaces, you could do:

```fish
string replace -rf " {$header_spaces[1]}" '' -- $results[5..]
```

The "f" stands for "filter" and means it will only let through lines where it replaced something. This should also remove the need for the `string sub` if I'm seeing this correctly?

Reply:
>  This should also remove the need for the `string sub` if I'm seeing this correctly?

No because that is also responsible to cut of the rest of the line.

Reply:
I only need the first column, and wouldn't that break on SSIDs containing whitespace?


And also the amount of leading whitespace is important because of line wrapping in some columns.

Reply:
This would work:

```fish
string match -rg "  .{$(math $header_spaces[1] - 2)}(.{$(math $header_non_spaces[1] + $header_spaces[2])})" $results[5..] | string trim
```

Reply:
> I only need the first column, and wouldn't that break on SSIDs containing whitespace?

Ah, forgot that that's a thing.

> And also the amount of leading whitespace is important because of line wrapping in some columns.

Hold on: Does iwctl wrap the lines *itself*? Because if the lines purely appear to wrap because of your terminal that's fine and there's no need to handle anything.

(if it does hard-wrap lines that's an extreme misfeature in iwctl)

Reply:
> it does, and it doesn't check if it is outputting to a terminal as usual. This also breaks awk IIRC

I can't see this.

With output like the above - from `iwctl device list`, I can resize my terminal to be extremely small and do `echo (iwctl device list)[5]`, and get *one line* that is too large for the terminal and appears to wrap. The second visual line doesn't appear to start with whitespace.

But it is one "physical" line, because the command substitution is split on newlines.

Hard-wrapping, on the other hand, would make 5 the first part of the line and 6 the continuation.

So I do not see how this is affected.

Note: Hard-wrapping would mean there would *always* be unavoidable edge-cases in parsing - what if the SSID has the right amount of whitespace in the right place so the second line appears to be aligned right?

Reply:
What the fudge. This program is cursed.

Reply:
@faho might be that it only happens for `adapter list` but it does. Haven't checked their code on how they do the wrapping.